libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+done:
+    return ret;
+}
+
+static cl_error_t insertPointer(PointerList *pl, const uint8_t *pointer)
+{
+    cl_error_t ret = CL_SUCCESS;
+
+    if (pl->cnt == (pl->capacity - 1)) {
+        uint32_t newCapacity = pl->capacity + POINTER_LIST_INCREMENT;
+        CLI_REALLOC(pl->idxs, newCapacity * sizeof(uint8_t *),
+                cli_errmsg("insertPointer: Can't allocate memory\n");
+                ret = CL_EMEM);
+
+        pl->capacity = newCapacity;
+    }
+
+    pl->idxs[pl->cnt++] = pointer;
+
+done:
+    return ret;
+}
+
+static cl_error_t findFileIdentifiers(const uint8_t *const input, PointerList *pfil)
+{
+
+    cl_error_t ret        = CL_SUCCESS;
+    const uint8_t *buffer = input;
+    uint16_t tagId        = getDescriptorTagId(buffer);
+
+    while (257 == tagId) {
+        if (CL_SUCCESS != (ret = insertPointer(pfil, buffer))) {
+            goto done;
+        }
+
+        buffer = buffer + getFileIdentifierDescriptorSize((FileIdentifierDescriptor *)buffer);
+        tagId  = getDescriptorTagId(buffer);
+    }
+
+done:
+    return ret;
+}
+
+static cl_error_t findFileEntries(const uint8_t *const input, PointerList *pfil)
+{
+
+    cl_error_t ret        = CL_SUCCESS;
+    const uint8_t *buffer = input;
+    uint16_t tagId        = getDescriptorTagId(buffer);
+
+    while (261 == tagId) {
+        if (CL_SUCCESS != (ret = insertPointer(pfil, buffer))) {
+            goto done;
+        }
+
+        buffer = buffer + getFileEntryDescriptorSize((FileEntryDescriptor *)buffer);
+        tagId  = getDescriptorTagId(buffer);
+    }
+
+done:
+    return ret;
+}
+
+cl_error_t cli_scanudf(cli_ctx *ctx, const size_t offset)
+{
+    cl_error_t ret                          = CL_SUCCESS;
+    size_t idx                              = offset;
+    size_t lastOffset                       = 0;
+    size_t i                                = 0;
+    uint8_t *buffer                         = NULL;
+    PrimaryVolumeDescriptor *pvd            = NULL;
+    GenericVolumeStructureDescriptor *gvsd  = NULL;
+    ImplementationUseVolumeDescriptor *iuvd = NULL;
+    LogicalVolumeDescriptor *lvd            = NULL;
+    PartitionDescriptor *pd                 = NULL;
+    UnallocatedSpaceDescriptor *usd         = NULL;
+    TerminatingDescriptor *td               = NULL;
+    LogicalVolumeIntegrityDescriptor *lvid  = NULL;
+    AnchorVolumeDescriptorPointer *avdp     = NULL;
+
+    bool isInitialized = false;
+    PointerList fileIdentifierList;
+    PointerList fileEntryList;
+
+    if (offset < 32768) {
+        return CL_SUCCESS; /* Need 16 sectors at least 2048 bytes long */
+    }
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, NUM_GENERIC_VOLUME_DESCRIPTORS * VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        ret = CL_SUCCESS;
+        goto done;
+    }
+
+    for (i = 0; i < NUM_GENERIC_VOLUME_DESCRIPTORS; i++) {
+        gvsd       = (GenericVolumeStructureDescriptor *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I'm a little concerned if this needs to be unneeded. Or maybe use `fmap_need_off_once()` instead